### PR TITLE
Implement __repr__ for route classes

### PR DESCRIPTION
**Pull request by alex-oleshkevich**
This little quality of life improvement makes debugging a bit easier by providing route information in the debugger:

image

### CODE REVIEW

**Comment by Kludex** [comment]
Are you strong about the style in this PR or can we do:

```
Route(path="/login", name="login", methods={"GET", "POST"})
```
?

**Comment by alex-oleshkevich** [comment]
> Are you strong about the style in this PR or can we do:
> 
> ```
> Route(path="/login", name="login", methods={"GET", "POST"})
> ```
> 
> ?

No, that's a matter of taste. If you think the proposed solution is more readable, I am fine to change.

**Review by Kludex — CHANGES_REQUESTED** [review]
My previous comment is not an opinion anymore, as I've noticed that we are already using the style I mentioned in our source code. Would you mind adapting it @alex-oleshkevich ?

File: starlette/routing.py [hunk 1]
```diff
@@ -440,6 +449,12 @@ def __eq__(self, other: typing.Any) -> bool:
             and self.app == other.app
         )
 
+    def __repr__(self) -> str:
+        return (
+            f"<{self.__class__.__name__}: "
+            f"path={self.path}, name={self.name or ''}, app={self.app}>"
```
Comment (Kludex) [inline starlette/routing.py]: 👍 
```suggestion
            f"path={self.path}, name={self.name or ''}, app={self.app!r}>"
```

Comment (Kludex) [reply starlette/routing.py]: Actually, the `!r` is used on all parameters in the other `__repr__` that we have in code 🤔

Comment (Kludex) [reply starlette/routing.py]: Ah, I got why... It's because with them, there's no need to wrap on parenthesis!

File: starlette/routing.py [hunk 2]
```diff
@@ -283,6 +283,13 @@ def __eq__(self, other: typing.Any) -> bool:
             and self.methods == other.methods
         )
 
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}("
+            f'path="{self.path}", name="{self.name}", '
+            f"methods={sorted(self.methods or [])})"
+        )
```
Comment (Kludex) [inline starlette/routing.py]:
```suggestion
        class_name = self.__class__.__name__
        methods = sorted(self.methods or [])
        path, name = self.path, self.name
        return f"{class_name}(path={path!r}, name={name!r}, methods={methods!r})"
```

File: starlette/routing.py [hunk 3]
```diff
@@ -341,6 +348,9 @@ def __eq__(self, other: typing.Any) -> bool:
             and self.endpoint == other.endpoint
         )
 
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}(path="{self.path}", name="{self.name}")'
```
Comment (Kludex) [inline starlette/routing.py]:
```suggestion
        return f"{self.__class__.__name__}(path={self.path!r}, name={self.name!r})"
```

File: starlette/routing.py [hunk 4]
```diff
@@ -440,6 +450,12 @@ def __eq__(self, other: typing.Any) -> bool:
             and self.app == other.app
         )
 
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}("
+            f'path="{self.path}", name="{self.name or ""}", app="{self.app!r}")'
+        )
```
Comment (Kludex) [inline starlette/routing.py]:
```suggestion
        class_name = self.__class__.__name__
        name = self.name or ""
        return f"{class_name}(path={self.path!r}, name={name!r}, app={self.app!r})"
```

File: starlette/routing.py [hunk 5]
```diff
@@ -507,6 +523,12 @@ def __eq__(self, other: typing.Any) -> bool:
             and self.app == other.app
         )
 
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}("
+            f'host="{self.host}", name="{self.name or ""}", app="{self.app!r}")'
+        )
```
Comment (Kludex) [inline starlette/routing.py]:
```suggestion
        class_name = self.__class__.__name__
        name = self.name or ""
        return f"{class_name}(host={self.host!r}, name={self.name!r}, app={self.app!r})"
```

File: starlette/routing.py [hunk 6]
```diff
@@ -847,3 +869,8 @@ def decorator(func: typing.Callable) -> typing.Callable:
             return func
 
         return decorator
+
+    def __repr__(self) -> str:
+        routes_count = len(self.routes)
+        noun = "route" if routes_count == 1 else "routes"
+        return f"{self.__class__.__name__}({routes_count} {noun})"
```
Comment (Kludex) [inline starlette/routing.py]: Why not the routes? 🤔

File: tests/test_routing.py [hunk 7]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
     resp = client.get("/mount/err")
     assert resp.status_code == 403, resp.content
     assert "X-Mounted" not in resp.headers
+
+
+def test_route_repr() -> None:
+    route = Route("/welcome", endpoint=homepage)
+    assert (
+        repr(route)
+        == "Route(path=\"/welcome\", name=\"homepage\", methods=['GET', 'HEAD'])"
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
        == "Route(path='/welcome', name='homepage', methods=['GET', 'HEAD'])"
```

File: tests/test_routing.py [hunk 8]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+    )
+
+
+def test_route_repr_without_methods() -> None:
+    route = Route("/welcome", endpoint=Endpoint, methods=None)
+    assert repr(route) == 'Route(path="/welcome", name="Endpoint", methods=[])'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "Route(path='/welcome', name='Endpoint', methods=[])"
```

File: tests/test_routing.py [hunk 9]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+
+
+def test_websocket_route_repr() -> None:
+    route = WebSocketRoute("/ws", endpoint=websocket_endpoint)
+    assert repr(route) == 'WebSocketRoute(path="/ws", name="websocket_endpoint")'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "WebSocketRoute(path='/ws', name='websocket_endpoint')"
```

File: tests/test_routing.py [hunk 10]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+
+
+def test_mount_repr() -> None:
+    route = Mount(
+        "/app",
+        routes=[
+            Route("/", endpoint=homepage),
+        ],
+    )
+    assert repr(route) == 'Mount(path="/app", name="", app="Router(1 route)")'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "Mount(path='/app', name='', app=\"Router(1 route)\")"
```

File: tests/test_routing.py [hunk 11]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+
+def test_mount_named_repr() -> None:
+    route = Mount(
+        "/app",
+        name="app",
+        routes=[
+            Route("/", endpoint=homepage),
+        ],
+    )
+    assert repr(route) == 'Mount(path="/app", name="app", app="Router(1 route)")'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "Mount(path='/app', name='app', app=\"Router(1 route)\")"
```

File: tests/test_routing.py [hunk 12]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+def test_host_repr() -> None:
+    route = Host(
+        "example.com",
+        app=Router(
+            [
+                Route("/", endpoint=homepage),
+            ]
+        ),
+    )
+    assert repr(route) == 'Host(host="example.com", name="", app="Router(1 route)")'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "Host(host='example.com', name='', app=\"Router(1 route)\")"
```

File: tests/test_routing.py [hunk 13]
```diff
@@ -952,3 +952,86 @@ async def modified_send(msg: Message) -> None:
...
+    route = Host(
+        "example.com",
+        name="app",
+        app=Router(
+            [
+                Route("/", endpoint=homepage),
+            ]
+        ),
+    )
+    assert repr(route) == 'Host(host="example.com", name="app", app="Router(1 route)")'
```
Comment (Kludex) [inline tests/test_routing.py]:
```suggestion
    assert repr(route) == "Host(host='example.com', name='app', app=\"Router(1 route)\")"
```

**Review by Kludex — COMMENTED** [review]
The comments are mainly to comply with the style we have on the other `__repr__`. 🙏 

I think the suggestions on the tests are correct, but not sure. 👀

File: starlette/routing.py (hunk 6 above)
Comment (alex-oleshkevich) [reply starlette/routing.py]: There may be many routes, and printing all of them is not helpful. Just some brief info about their count, IMO, is enough.

Comment (Kludex) [reply starlette/routing.py]: But then... Is it really helpful to have just the number of routes? :thinking:

**Review by Kludex — APPROVED** [review]
I'm not sure about the `Router.__repr__`, as I don't see much value there, but I'll let you judge that.

Let me know your judgement, and I'll merge it. :+1: 

Thanks for making an effort to improve our users' lives @alex-oleshkevich :pray:

**Comment by alex-oleshkevich** [comment]
> I'm not sure about the `Router.__repr__`, as I don't see much value there, but I'll let you judge that.
> 
> Let me know your judgement, and I'll merge it. +1
> 
> Thanks for making an effort to improve our users' lives @alex-oleshkevich pray

okay, i will remove it. this is too confusing.

**Review by Kludex — DISMISSED** [review]

**Comment by Kludex** [comment]
image

I'm not able to merge it :sweat_smile:

**Comment by Kludex** [comment]
Does anyone know why I can't merge? 🤔

**Comment by alex-oleshkevich** [comment]
@Kludex what's behind "statuses" link?

**Comment by Kludex** [comment]
> @Kludex what's behind "statuses" link?

https://docs.github.com/rest/commits/statuses

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
I don't know :+1:

**Comment by Kludex** [comment]
The target is wrong... It should be "master", but it's "encode:master"... I don't know what happened here.

image

**Comment by Kludex** [comment]
Easy. :eyes:

**Pull request merged**